File: BI-TJV/semestral/rest-client/app/src/main/java/cz/wrzecond/restsys/employee/EmployeeAdapter.java
package cz.wrzecond.restsys.employee;

import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AlertDialog;

import java.util.List;
import java.util.Locale;

import cz.wrzecond.restsys.R;
import cz.wrzecond.restsys.task.HttpMethod;
import cz.wrzecond.restsys.task.HttpRequest;
import cz.wrzecond.restsys.Activity;
import cz.wrzecond.restsys.Adapter;
import wrzecond.employee.TjvEmployeeReadDTO;

public class EmployeeAdapter extends Adapter<TjvEmployeeReadDTO> {

    // Constructors
    public EmployeeAdapter (Activity<TjvEmployeeReadDTO> activity, @NonNull List<TjvEmployeeReadDTO> dtos) {
        super (activity, R.layout.item_employee, dtos);
    }
    public EmployeeAdapter (Activity<TjvEmployeeReadDTO> activity, @StringRes int errorResId) {
        super (activity, R.layout.item_employee, errorResId);
    }

    @Override
    protected Adapter<TjvEmployeeReadDTO>.ViewHolder createViewHolder (View view) {
        return new EmployeeAdapter.ViewHolder(view);
    }

    protected class ViewHolder extends Adapter<TjvEmployeeReadDTO>.ViewHolder {

        // === Properties
        private TextView employeeUsername;
        private TextView employeeName;

        private ImageView employeeEdit;
        private ImageView employeeDelete;

        public ViewHolder (@NonNull View view) { super(view); }

        @Override
        protected void initView (@NonNull View view) {
            this.employeeUsername = view.findViewById(R.id.employee_username);
            this.employeeName = view.findViewById(R.id.employee_name);
            this.employeeEdit = view.findViewById(R.id.employee_edit);
            this.employeeDelete = view.findViewById(R.id.employee_delete);
        }

        @Override
        protected void update ( TjvEmployeeReadDTO dto ) {
            employeeUsername.setText(dto.getUsername());
            employeeName.setText(String.format(Locale.getDefault(), "%s %s",
                    dto.getFirstName(), dto.getLastName()));

            // Only for admins
            if ( !activity.getService().isAdmin() ) {
                employeeEdit.setVisibility(View.GONE);
                employeeDelete.setVisibility(View.INVISIBLE);
                return;
            }

            employeeEdit.setOnClickListener( v -> {
                AlertDialog dialog = new AlertDialog.Builder(activity)
                    .setView(R.layout.create_employee)
                    .create();

                dialog.setOnShowListener(activity.createListener(dialog, dto));
                dialog.show();
            } );

            // Do not allow self-delete
            if (dto.getId() == activity.getService().getLoggedUserID()) {
                employeeDelete.setVisibility(View.INVISIBLE);
                return;
            }

            employeeDelete.setOnClickListener( v ->
                new AlertDialog.Builder(activity)
                    .setTitle(R.string.employee_delete)
                    .setMessage(activity.getString(R.string.employee_delete_confirm, dto.getFirstName(), dto.getLastName()))
                    .setNegativeButton(R.string.button_dismiss, null)
                    .setPositiveButton(R.string.button_delete, ((dialog, which) -> {
                        // Schedule request
                        HttpRequest request = new HttpRequest("/employees/" + dto.getId(), HttpMethod.DELETE);
                        activity.getService().scheduleTask( (response) -> {
                            // Show message and close
                            Toast.makeText(activity,
                                    response.getStatus().isSuccess() ? activity.getString(R.string.delete_success) :
                                        activity.getString(R.string.delete_failure, response.getStatus().name()),
                                    Toast.LENGTH_SHORT).show();

                            // Hide and reload list if success
                            dialog.dismiss();
                            if (response.getStatus().isSuccess())
                                activity.loadList();
                        }, request );
                    }))
                    .create().show()
            );
        }

    }

}

File: BI-KOT/semestral/rest-client/app/src/main/java/cz/wrzecond/restsys/employee/EmployeeAdapter.kt
package cz.wrzecond.restsys.employee

import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import cz.wrzecond.restsys.Adapter
import cz.wrzecond.restsys.R
import cz.wrzecond.restsys.databinding.ItemEmployeeBinding
import cz.wrzecond.restsys.task.HttpMethod
import cz.wrzecond.restsys.task.HttpRequest
import wrzecond.TjvEmployeeReadDTO
import java.util.*

class EmployeeAdapter (activity: EmployeeActivity, dtos: List<TjvEmployeeReadDTO>? = null, errorResId: Int? = null)
: Adapter<TjvEmployeeReadDTO>(activity, dtos, errorResId) {

    override fun createViewHolder (parent: ViewGroup)
        = ViewHolder(ItemEmployeeBinding.inflate(activity.layoutInflater, parent, false))

    inner class ViewHolder (private val binding: ItemEmployeeBinding) : Adapter<TjvEmployeeReadDTO>.ViewHolder(binding.root) {
        override fun update (dto: TjvEmployeeReadDTO) {
            // Show name
            binding.employeeName.text = String.format(Locale.getDefault(), "%s %s", dto.firstName, dto.lastName)

            // Only for admin
            if (!activity.service.isAdmin) {
                binding.employeeEdit.visibility = View.GONE
                binding.employeeDelete.visibility = View.GONE
            }

            // Edit employee
            binding.employeeEdit.setOnClickListener {
                AlertDialog.Builder(activity)
                    .setView(R.layout.create_employee)
                    .create()
                    .apply { setOnShowListener(activity.createListener(this, dto)) }
                    .show()
            }

            // No self-delete
            if (dto.id == activity.service.loggedUserID) {
                binding.employeeDelete.visibility = View.INVISIBLE
                return
            }

            // Delete employee
            binding.employeeDelete.setOnClickListener {
                AlertDialog.Builder(activity)
                    .setTitle(R.string.employee_delete)
                    .setMessage(activity.getString(R.string.employee_delete_confirm, dto.firstName, dto.lastName))
                    .setNegativeButton(R.string.button_dismiss, null)
                    .setPositiveButton(R.string.button_delete) { di,_ ->
                        val request = HttpRequest("/employees/${dto.id}", HttpMethod.DELETE)
                        activity.service.scheduleTask(request) { response ->
                            Toast.makeText(activity, response.getToastString(activity), Toast.LENGTH_SHORT).show()
                            di.dismiss()
                            if (response.status.isSuccess)
                                activity.loadList()
                        }
                    }
                    .create()
                    .show()
            }
        }
    }

}